Extension cases sync

proteusvacuum opened the pull request:
Opening for tests / review. This is still in the preliminary stages...
@czue would be great if you could give this a glance (should be easiest commit by commit)

Todo:
- [ ] Fix skipped tests in `.json` shared test file
- [ ] Test subsequent syncs still do the right thing (e.g. delete, delegate, close, etc. still sends the correct cases)
- [ ] Make sure ownership cleanliness gets set properly with extensions

proteusvacuum committed:
Update case/related view and add dbaccessor

The view now emits [domain, referenced_id, "reverse_index",
relationship]

proteusvacuum committed:
Include extensions in case footprint

proteusvacuum committed:
Add extension_index_tree to SimplifiedSyncLog

proteusvacuum committed:
Add get_all_dependencies to IndexTree

proteusvacuum committed:
Add extension indices during sync

proteusvacuum committed:
For clean sync, base set includes extensions

proteusvacuum committed:
Test get_all_dependencies

proteusvacuum committed:
Test prune_case

proteusvacuum committed:
Test SyncTokenUpdates

proteusvacuum committed:
Add shared mobile and HQ tests

updates sniffer to reload when json files are saved

dimagimon commented on corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py:
```diff
@@ -20,6 +20,7 @@
 from casexml.apps.phone.restore import CachedResponse, RestoreConfig, RestoreParams, RestoreCacheSettings
 from casexml.apps.case.xml import V2, V1
 from casexml.apps.case.sharedmodels import CommCareCaseIndex
+from casexml.apps.case.const import CASE_INDEX_EXTENSION
```
F401 'CASE_INDEX_EXTENSION' imported but unused

proteusvacuum commented on corehq/ex-submodules/casexml/apps/phone/models.py:
```diff
@@ -472,6 +473,41 @@ class IndexTree(DocumentSchema):
         return json.dumps(self.indices, indent=2)
 
+    @staticmethod
+    def get_all_dependencies(case_id, child_index_tree, extension_index_tree,
```
This is the crux of it. Still not convinced that this completely covers the spec (cf. skipped tests).

czue commented:
wow, this is a beast! looks great overall and love all the tests. I didn't review those too closely - let me know if you think I should.

proteusvacuum committed:
Don't make case dependent if ther are child indices

proteusvacuum committed:
Include all index identifiers

proteusvacuum committed:
Add extension with 2 hosts

proteusvacuum committed:
Refactor -> extension updates aren't always live

Makes it clearer that some extension updates aren't live, but are still
added to the case ids on the phone

proteusvacuum committed:
Refactor -> pull out delete_index

proteusvacuum committed:
Dont sync purged cases that the phone never knew about

proteusvacuum committed:
Fix test

proteusvacuum committed:
get dirtiness flags for all outgoing indices

proteusvacuum committed:
More complex extension chain cleanliness

proteusvacuum committed:
Add branching test

proteusvacuum committed:
Add changing ownership tests

proteusvacuum committed:
Verify cleanliness flags

proteusvacuum committed:
Get all reverse indices also returns extensions

proteusvacuum committed:
get wrapped extension cases

proteusvacuum committed:
Hints for extensions are the extensions

proteusvacuum committed:
Generic set cleanliness flags

proteusvacuum committed:
Test owned extensions

proteusvacuum committed:
Get dependent cases from db

proteusvacuum committed:
Update cleanliness test with new hint defs

proteusvacuum committed:
Get extension cleanliness with hint setting

proteusvacuum committed:
Hint is the referenced case when looking at indices

proteusvacuum committed:
More test cases

proteusvacuum committed:
WIP hint invalidation

proteusvacuum committed:
Sniffer only run tests once

proteusvacuum committed:
Beautify + add closed extension test

proteusvacuum committed:
Properly set properties on indexes in tests

proteusvacuum committed:
New Purge Algorithm

proteusvacuum committed:
Remove decorator

proteusvacuum committed:
Fix test outcome

proteusvacuum committed:
Properly set dependent case ids

proteusvacuum committed:
Add closed cases on sync

proteusvacuum committed:
purge indices of removed cases

proteusvacuum committed:
Exit infinite loop

proteusvacuum committed:
allow very circular loops

proteusvacuum committed:
don't remove closed cases from dependents

proteusvacuum committed:
prune -> purge

proteusvacuum committed:
Refactor

proteusvacuum committed:
more refactor

proteusvacuum committed:
bad merge

dimagimon commented:
There are 9 errors:

* corehq/ex-submodules/casexml/apps/case/util.py, line 131 - E302 expected 2 blank lines, found 1
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 4 - F401 'CASE_INDEX_EXTENSION' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 5 - F401 'get_reverse_indices_for_case_id' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 9 - F401 'get_extension_cases' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 9 - F401 'get_indexed_cases' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 134 - W291 trailing whitespace
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 146 - E501 line too long (131 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/tests/test_cleanliness.py, line 642 - E501 line too long (118 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/tests/test_extension_indexes.py, line 105 - E501 line too long (117 > 115 characters)

proteusvacuum commented:
(just wanted to run tests again, will clean up history and ping when ready for review again)

dimagimon commented:
There are 9 errors:

* corehq/ex-submodules/casexml/apps/case/util.py, line 131 - E302 expected 2 blank lines, found 1
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 4 - F401 'CASE_INDEX_EXTENSION' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 5 - F401 'get_reverse_indices_for_case_id' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 9 - F401 'get_extension_cases' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 9 - F401 'get_indexed_cases' imported but unused
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 134 - W291 trailing whitespace
* corehq/ex-submodules/casexml/apps/phone/cleanliness.py, line 146 - E501 line too long (131 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/tests/test_cleanliness.py, line 642 - E501 line too long (118 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/tests/test_extension_indexes.py, line 105 - E501 line too long (117 > 115 characters)
